### ISSUE TITLE
 value is not compatible with the variable's type constraint: all map elements must have the same type

### ISSUE THREAD

**Issue by aniro** [hashicorp/terraform#21384]
### Terraform Version

```
Terraform v0.12.0-rc1 on Windows 10 x64
```

### Terraform Configuration Files

```hcl
variable "test" {
    type = "map"
    default = {
        a = {
            x = 1
            y = true
            }
        b = {
            z = 2
            }
    }
}

output "test_out" {
  value = "${var.test}"
}
```

### Expected Behavior
```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

test_out = {
  "a" = {
    "x" = 1
    "y" = true
  }
  "b" = {
    "z" = 2
  }
}
```

### Actual Behavior
```
Error: Invalid default value for variable

  on main.tf line 3, in variable "test":
   3:     default = {
   4:         a = {
   5:             x = 1
   6:             y = true
   7:             }
   8:         b = {
   9:             w = 2
  10:             }
  11:     }

This default value is not compatible with the variable's type constraint: all
map elements must have the same type.
```

### Steps to Reproduce
1. `terraform apply`

### Additional Context
Any of following actions fix this issue:
1. removing `type = "map"` from variable definition
2. putting `true` value in quotes: `y = "true"`
3. adding second inner element to map element `b`:
```
        b = {
            z = 2
            q = "3"
            }
```

**Comment by apparentlymart** [comment]
Hi @aniro! Thanks for reporting this.

When you say `list` in Terraform 0.12, that triggers a compatibility behavior where Terraform will try to find the closest new-system type that will match what is in `default`. In this case, it seems to be getting tripped up because the maps contain a mixture of numbers and bools and so it can't decide whether to use `list(map(number))` or `list(map(bool))` -- neither of those would work here, because numbers and booleans are not type-compatible.

As you saw, introducing at least one string into the maps gives Terraform a new option: `list(map(string))`. Because numbers and bools can both convert to string, Terraform takes that opportunity and produces a list of maps of strings instead:

```
test_out = {
  "a" = {
    "x" = "1"
    "y" = "true"
  }
  "b" = {
    "z" = "2"
  }
}
```

If those values being converted to strings is acceptable -- it _should_ be in most cases, since Terraform will generally auto-convert those values back to numbers and bools again when necessary -- then changing the type explicitly to `list(map(string))` will tell Terraform how to resolve this, and then the manual quoting of the values won't be necessary.

A further option is to simply set the type to `any`, which means what it sounds like: Terraform won't apply any type constraints or conversions at all, and will just let your data structure pass through exactly as you wrote it.

We have some additional notes on this in the final draft of the upgrade guide. It's not published on the site yet since it's staged up for the v0.12.0 final release, but you can read it on GitHub in the mean time if you like: [Type Constraints on Variables](https://github.com/hashicorp/terraform/blob/233115ad4ba06874b781aaa2056adfa3838089b7/website/upgrade-guides/0-12.html.markdown#type-constraints-on-variables). The error message in your case is a little different because of the way Terraform interpreted the default value, but the same root cause applies. (The upgrade tool mentioned in the guide would've rewritten this to `type = list(string)`, leading to the error message shown in the guide.)

**Comment by aniro** [comment]
Thanks @apparentlymart for the detailed explanation. It was not evident that `map` actually translates to `map(map(string))`. I would expect `map(any)` instead which seems more logical.

And this actually does not explain the following case:
main.tf:
```
variable "test" {
    type = "map"
    default = {}
}

output "test_out" {
  value = "${var.test}"
}
```
It should not have guessed anything weird from default since default is just an empty map.
But, providing the following var
terraform.tfvars:
```
test = {
        a = {
            x = 1
            y = true
            }
        b = {
            z = 2
            }
    }
```
also fail validation:
```
Error: Invalid value for input variable

  on terraform.tfvars line 1:
   1: test = {
   2:         a = {
   3:             x = 1
   4:             y = true
   5:             }
   6:         b = {
   7:             z = 2
   8:             }
   9:     }

The given value is not valid for variable "test": all map elements must have
the same type.
```
I believe this behavior needs some changes to be more predictable or at least be documented better / provide more clear error message.

**Comment by apparentlymart** [comment]
Hi @aniro,

`map(any)` means "map of any _single_ type". That is, you can use whatever type you like but all of the elements must have the _same_ type. That is what the error message is trying to explain here.

In order to have differently-typed elements, what you would need to use instead is [an object type](https://www.terraform.io/docs/configuration/types.html#object-), which allows you to tell Terraform the type for each attribute individually, rather than `map` where all of the elements must have the same type. 

```hcl
variable "test" {
  type = object({
    x = number
    y = bool
    z = number
  })
  default = {}
}
```

One way or another, Terraform must know the type of each attribute in order to check that the given value is correct. If the type checking is not desirable (if you just want to provide an arbitrary data structure with any mixture of types) then that is what the `any` constraint is for, and why I suggested that as the fallback if you don't want to tell Terraform exactly what type is needed here. `any` is, essentially, an opt-out mechanism for situations where a module was relying on the general lack of type checking for variables in Terraform 0.11, which applies to the examples shown here.

**Comment by aniro** [comment]
Hi @apparentlymart, 
> map(any) means "map of any single type". That is, you can use whatever type you like but all of the elements must have the same type.
In my example, all elements of variable do have the same type which is map (map of maps). The actual correct constraint for my case would be map(map(any))
So it is still not clear why maps as the elements of map is an error. The error message is misleading.

Moreover if I use `type = map(map(any))` it still does not work, failing with the same error.

BTW, I switched to release version, and trying the following causes panic (not sure if it was the case on rc1):
```
variable "test" {
    type = map(map(any)) <--- changed
    default = {
        a = {
            x = 1
            y = "true"  <--- quoting 
            },
        b = {
            z = 2
            }
    }
}

output "test_out" {
  value = "${var.test}"
}
```
https://gist.github.com/aniro/23a46d9549104f42948495a083756dc4

**Comment by apparentlymart** [comment]
Your map of maps includes both numbers and booleans, which are not the same type.

Whenever you see `any` in a type constraint, it means that Terraform will try to find a specific type to replace `any` and thus produce a concrete type, rather than just a type _constraint_. Therefore if you write `map(any)` Terraform must try to find a single type to replace `any` with that all of the element values can convert to.

If all of your map elements were numbers, then Terraform would choose `map(map(number))`. If they were all booleans, it would be `map(map(bool))`. But if you have both numbers and booleans then there is no single type that `any` could be replaced with, because `map(map(number))` would fail to accept `true` and `map(map(bool))` would fail to accept `1` or `2`.

The reason that `type = any` works is because Terraform is then able to replace `any` with a very specific structural type which the value conforms to:

```hcl
object({
  a = object({
    x = number
    y = bool
  }),
  b = object({
    z = number
  }),
})
```

The reason that `map(map(string))` works is that numbers and bools can both convert to string, as I described in an earlier comment.

If you want to have mixed attribute types and you don't want to define an explicit object type then `type = any` is the only way to do that, because it allows Terraform to derive a very specific object type like the above automatically, and thus meet the type constraint. For a new configuration written from scratch I would recommend using a map of a specific object type, but that is likely to be a breaking change for your module's callers, and so `any` will preserve compatibility with existing callers.

The panic when you use `map(map(any))` is certainly a bug, though. It should have been a similar error than what you got before, but now one level deeper in the structure:

```
This default value is not compatible with the variable's type constraint:
element ["a"]: all map elements must have the same type.
```

**Comment by aniro** [comment]
Thanks, I understand how it works, now. What I am trying to say is that current behavior does not seem intuitive and error message it produces along with documentation do not help in understanding of what is going on. 
Specifically the following makes a lot of confusion:
1. `map(any)` - is not really **any**
2. `map(any)` goes all the way down in variable structure and maps all leaf elements to guessed static types. While I would expect it to cover only the first level of  nesting. I.e. all elements of the upper level map should be of the same type irrespective of what is inside of inner maps.
3. There seem to be now way (please correct me if I am wrong) to tell terraform that my var should be of type `map(any)` or `list` or `list(map(any))` and we do not care about what is inside values of map or list. For instance there can be some optional parameters like this:
```
disks = [ 
  { size = 20, thin_provisioned = false}, 
  { size = 30 }
]
```
it is not possible to define type constraint for the example above and one have to use `any` instead of `list(map(any))`:
I could use `list(map(string))` which would work, but with type conversion to strings which is not the best approach as to me.
4. The way how quotation or addition of another string element helps to magically change the behavior of `map(any)` type leading to implicit type conversions.

**Comment by apparentlymart** [comment]
Thanks for the feedback, @aniro! Once we've dealt with the initial set of bug fixes from the 0.12.0 release I'll see about revising the Type Constraints documentation to explain this more explicitly.

The key thing here is that `any` is not actually a _type_, but rather a type _constraint_. The current documentation is not particularly clear on that, but Terraform must know the concrete type of every value in order to complete type checking, because Terraform must deal with lots of "unknown" values during the planning phase but still needs to know what types those values have. This information not being available in prior versions of Terraform was the cause of lots of strange bugs, and that is why Terraform is now designed to require precise type information.

You're correct that there is not a way to support optional attributes in an object type: they must always all be set. You can set them to `null` to explicitly omit them, but because variable values are values -- as opposed to nested blocks, as we see in some `resource` blocks -- the idea of "optional" doesn't exist for them. Perhaps in a future release we'll introduce a way for modules to define both arguments and nested blocks for their configuration, so that their configurations can be built from a similar featureset as for resource types. 

I'm going to relabel this issue to indicate that it represents documentation work. While we're there we can also see if there's a way to improve the error message; it already contains a correct description of the problem, but it doesn't give any information on what one might do to resolve it.

**Comment by smiller171** [comment]
I think it would fix a lot of problems if you could define types for complex objects. The syntax you showed above for the actual types is basically what I'm thinking here.

```
object({
  a = object({
    x = number
    y = bool
  }),
  b = object({
    z = number
  }),
})
```
